### ISSUE TITLE
CPU Limits: a Che 7.9 workspace fails to start with Che 7.13

### ISSUE THREAD

**Issue by l0rd** [eclipse-che/che#16975]
### Describe the bug

I have tried to run a [java-mysql](https://gist.github.com/l0rd/c8277e42c92a8a76370ca5ead663d010#file-devfile-yaml) workspace on a couple of OpenShift clusters:
1. RHDPS opentlc where users namespaces have some default resource limits
2. OpenShift devcluster where users namespace do NOT have default resource limits
and starting the workspace failed in both cases.

In the first case I got the error
```
Error creating: pods "workspacentxzhlu35yvfn4c6.workspace-5bbcb9cb9f-k6h2q" is forbidden: maximum cpu usage per Pod is 4, but limit is 9'
```

On the second one the pod fails to get scheduled because there is no node able to run it.

If I use CRW 2.1.1, and [the same devfile](https://gist.github.com/l0rd/c8277e42c92a8a76370ca5ead663d010#file-devfile-yaml), the workspace is started successfully on both clusters.

### Che version

nightly

### Steps to reproduce

Deploy Che on OCP 4.3 and start a workspace using a factory link referencing [this devfile](https://gist.github.com/l0rd/c8277e42c92a8a76370ca5ead663d010#file-devfile-yaml).

### Expected behavior

The workspace starts

### Runtime

Openshift (include output of `oc version`)

```bash
$ oc version
Client Version: version.Info{Major:"4", Minor:"2+", GitVersion:"v4.2.0", GitCommit:"61fc89fab", GitTreeState:"clean", BuildDate:"2019-08-28T22:30:54Z", GoVersion:"go1.12.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"b3bfb5a", GitTreeState:"clean", BuildDate:"2020-03-02T08:50:52Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
OpenShift Version: 4.3.5
```

### Screenshots

![image](https://user-images.githubusercontent.com/606959/82384673-a379d600-9a30-11ea-80d3-08909756f756.png)

### Installation method

```bash
$ chectl version
chectl/0.0.20200515-next.96decfd darwin-x64 node-v13.2.0
$ chectl server:start --installer=operator \
                    --platform=openshift \
                    --os-oauth \
                    --self-signed-cert \
                    -n che-server
```

### Environment

On both opentlc.com (RHDPS) and devcluster.openshift.com.

### Additional context

Workspace containers resources on Che 7.13  (the pod fails to be start)
```json
{
  "name": "che-jwtproxygnixa2cq",
  "cpu_lim": "2",
  "cpu_req": "125m",
  "mem_lim": "134217728",
  "mem_req": "134217728"
}
{
  "name": "mysql",
  "cpu_lim": "2",
  "cpu_req": "125m",
  "mem_lim": "314572800",
  "mem_req": "209715200"
}
{
  "name": "tools",
  "cpu_lim": "2",
  "cpu_req": "125m",
  "mem_lim": "1048576k",
  "mem_req": "209715200"
}
{
  "name": "vscode-javaatf",
  "cpu_lim": "1",
  "cpu_req": "115m",
  "mem_lim": "1572864k",
  "mem_req": "67108864"
}
{
  "name": "che-machine-execnuk",
  "cpu_lim": "1",
  "cpu_req": "115m",
  "mem_lim": "134217728",
  "mem_req": "67108864"
}
{
  "name": "theia-idejfc",
  "cpu_lim": "1",
  "cpu_req": "115m",
  "mem_lim": "512M",
  "mem_req": "67108864"
}
```

The same containers with CRW 2.1 (the pod is scheduled successfully)
```json
{
  "name": "che-jwtproxysxt360r4",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "134217728",
  "mem_req": "134217728"
}
{
  "name": "mysql",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "314572800",
  "mem_req": "314572800"
}
{
  "name": "tools",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "1048576k",
  "mem_req": "1048576k"
}
{
  "name": "vscode-java9bf",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "1572864k",
  "mem_req": "1572864k"
}
{
  "name": "che-machine-execueq",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "134217728",
  "mem_req": "134217728"
}
{
  "name": "theia-idel9q",
  "cpu_lim": null,
  "cpu_req": null,
  "mem_lim": "512M",
  "mem_req": "512M"
}
```

**Comment by l0rd** [comment]
https://github.com/eclipse/che/issues/16685 may address this problem for getting started projects.

But if we want to avoid that a migration from 7.9 to 7.14 (CRW 2.1 to 2.2) breaks existing workspaces on OpenShift clusters I am afraid that we should avoid setting CPU/Memory limits if they are not explicitly specified in the plugins meta.yaml and devfiles.

cc @skabashnyuk @ericwill

**Comment by mshaposhnik** [comment]
Hello, what i see here:
"jwtproxy", "tools", "mysql" - k8s and dockerimage tools comes from non-plugins default configuration (`che.workspace.default_cpu_limit_cores=2`). They all except internal tools (jwtproxy) can be overriden in devfile.
"theia", "che-machine-exec", "vscode-java" - came from plugin registry or use plugins default `che.workspace.sidecar.default_cpu_limit_cores=1`. They all can be overridden either on devfile or meta.yaml

Maybe it's make sense to lower down default for non-plugins to, say, 0.5?  It's quite enought for jwtproxy, while other machines can request more via devfile.

**Comment by skabashnyuk** [comment]
I think that `che-machine-exec" ` should be enough 0.5 cpu too.

**Comment by l0rd** [comment]
@mshaposhnik @skabashnyuk if I had created my workspace with Che 7.9 and I upgrade to Che 7.13 I expect my workspace to work (without manually edit the devfile). 

Lowering down the defaults can be a workaround yes. But I am wondering if, to stay on the safe side, we should  avoid applying defaults in wsmaster: it's the responsibility of the plugin and devfile maintainers to specify the cpu/mem requirements for the corresponding containers. If there are not CPU/MEM limits/requests we should not set them automatically because the side effect may be that the workspace won't start and that's the worst thing that can happen.

**Comment by skabashnyuk** [comment]
@l0rd  don't you consider as a problem if we don't set default limits and devfile authors will not set them too - as a result, CPU intensive operations in one workspace can affect other workspaces allocated at the same node? Setting some defaults for CPU limits can obligate devfile authors to be careful with the values that they going to set.

**Comment by l0rd** [comment]
> @l0rd don't you consider as a problem if we don't set default limits and devfile authors will not set them too - as a result, CPU intensive operations in one workspace can affect other workspaces allocated at the same node?

This is a good question. If we do not set resources automatically the workspace will have unlimited access to the node resources (if the namespace doesn't have some default limits: devcluster) or will be assigned some arbitrary default and potentially wrong values (if the namespace has some default limits: rhdps). On the other end, if we set automatically the resources of every arbitrary container, we will be wrong at some point and the workspace won't start, a workspace that would have worked perfectly without any automatic resource settings. So both solutions have their drawbacks but automatically enforcing some resources spec that will break a workspace is imo worse.

In other words we should be responsible to set the resources of the containers that we maintain (plugins, and getting started devfiles), we should allow user to configure their workspaces resources but we should not take the extra-responsibility to fix user containers because we cannot guarantee that we won't break their workspaces. That's also the Kubernetes approach: it doesn't automatically set the containers resources but let users do it.

**Comment by mshaposhnik** [comment]
@lord Should we also deperecate default memory limits?

**Comment by l0rd** [comment]
> @lord Should we also deperecate default memory limits?

Yes but to avoid suprises in the 7.9 -> 7.14 migration we should first make sure that our plugins and devfiles have their memory limit/request specified.

**Comment by l0rd** [comment]
cc @alexeykazakov

**Comment by alexeykazakov** [comment]
Here are my thoughts about the limits/requests.

I would do the following:
- Allow setting up limits/request in dev file (not sure if the existing getting-started dev files should have them though).
- If not set in dev file then use the optional default values from Che Server configuration.
- If the default values are not set then do not set limits. Admins will be able to enforce their own limits using LimitRanges (but it would be for all containers in the pod).

Other important things:
- Container limits/request configs/params should be container type specific (for all container types in all pods created in workspace namespace, not only for containers in workspace pods). Different containers hav different resource requirements. For example for jstproxy it could be something like this: `CHE_WORKSPACE_DEFAULT_JWTPROXY_CPU__LIMIT__CORES: '0.1'. 
- Do not ask more than you need! For example I don't know what `jwtproxy` container does but does it really require up to 2 (!) cores? Remember that those containers run in user namespaces and eat up namespace/user quotas.

**Comment by l0rd** [comment]
> Allow setting up limits/request in dev file (not sure if the existing getting-started dev files should have them though).
If not set in dev file then use the optional default values from Che Server configuration.

That's exactly the current situation: limits/requests can be specified in devfiles and in the plugins definitions. But getting started devfiles do not have any CPU limits/requests. And Che server enforce some limits/request default values if those are not specified in the devfiles or plugins definition.

And current situation is not satisfying: workspaces fails to start because of wrong default values. You should know that well.

> If the default values are not set then do not set limits. Admins will be able to enforce their own limits using LimitRanges (but it would be for all containers in the pod).

What do you mean? Are you suggesting to have a Che server property that controls if the limits will be enforced or not?

> Container limits/request configs/params should be container type specific (for all container types in all pods created in workspace namespace, not only for containers in workspace pods). Different containers hav different resource requirements. For example for jstproxy it could be something like this: `CHE_WORKSPACE_DEFAULT_JWTPROXY_CPU__LIMIT__CORES: '0.1'.
Do not ask more than you need! For example I don't know what jwtproxy container does but does it really require up to 2 (!) cores? Remember that those containers run in user namespaces and eat up namespace/user quotas.

You are right. We should set a lower CPU limit to the jwtproxy container. But that's not enough. The workspace sample in this issue description would still fail to start (the max CPU limit is 4 and my workspace would have CPU limit of 7). And the same workspace starts successfully on an older version of Che.

Anyway these are currently the defaults:

Non-pugin or editor type machines (dev-containers, jwtproxy, kubernetes / dockerimage type components etc):

Default RAM Limit: 1024 Mb (existing and unchanged)
Default RAM request: 200 Mb
Default CPU limit: 2 Cores
Default CPU request: 0.125 Cores

Plugin or editor type machines (can be overridden in plugin meta.yaml or devfile* ) :

Default RAM Limit: 128 Mb (existing and unchanged)
Default RAM request: 64 Mb
Default CPU limit: 1 Cores
Default CPU request: 0.115 Cores

**Comment by alexeykazakov** [comment]
> What do you mean? Are you suggesting to have a Che server property that controls if the limits will be enforced or not?

I'm suggesting do not set limits at all if there is no limits in devfile AND the default limits are not explicitly set via a Che server property. No any hidden (from users) default values.
So, if I don't specify the limits in CheCluster resource for example and don't have them in my devfile then no limits are set by Che. I would still able to set limits by creating my own LimitRange in the workspace namespace if I want.

> You are right. We should set a lower CPU limit to the jwtproxy container. But that's not enough. The workspace sample in this issue description would still fail to start (the max CPU limit is 4 and my workspace would have CPU limit of 7). And the same workspace starts successfully on an older version of Che.

The problem with the new Che is not that you set the defaults. The problem is that the defaults are too big. I still believe that Che admins should be able to set global default limits because Che creates multiple containers with different requirements and using LimitRanges won't work well in most cases. But the limits should be reasonable.
I would say that a regular workspace should not ask for more than 2 cores in total by default. So spread the default values among the containers so, for example IDE itself gets most of those 2 cores and tiny light stuff like jwtproxy uses not more then tiny fraction of those 2 cores.
If there are special needs for some specific workspaces/devfiles then keep those limits in the corresponding devfile then.
